File: Euclid/Game/Pathfinder/Position.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Euclid.Game.Entity.Humanoid;

namespace Euclid.Game
{
    public class Position
    {
        public int X;
        public int Y;
        public double Z;
        public int BodyRotation;
        public int HeadRotation;

        public int Rotation
        {
            get
            {
                return BodyRotation;
            }
            set
            {
                BodyRotation = value;
                HeadRotation = value;
            }
        }

        public Position() : this(0, 0, 0)
        {
        }

        public Position(int x, int y)
        {
            X = x;
            Y = y;
            Z = 0;
        }

        public Position(int x, int y, double z)
        {
            X = x;
            Y = y;
            Z = z;
        }

        public Position(int x, int y, double z, int headRotation, int bodyRotation)
        {
            X = x;
            Y = y;
            Z = z;
            HeadRotation = headRotation;
            BodyRotation = bodyRotation;
        }

        /// <summary>
        /// Get current tile by supplying room instance
        /// </summary>
        public RoomTile GetTile(Room room)
        {
            if (room == null || room.Model == null || !room.Model.IsTile(this))
                return null;

            if (room.Model.TileStates[this.X, this.Y] == TileState.CLOSED)
                return null;


            return room.Mapping.Tiles[X, Y];
        }

        /// <summary>
        /// Get if the position is adjacent to this position
        /// </summary>
        public bool Touches(Position position)
        {
            return GetDistanceSquared(position) <= 2;
        }


        /// <summary>
        /// Add coordinates to current position
        /// </summary>
        public Position Add(Position other)
        {
            return new Position(other.X + X, other.Y + Y, other.Z + Z);
        }

        /// <summary>
        /// Subtract coordinates to current position
        /// </summary>
        public Position Subtract(Position other)
        {
            return new Position(other.X - X, other.Y - Y, other.Z - Z);
        }

        /// <summary>
        /// Get the closest tile to this current tile
        /// </summary>
        public Position ClosestTile(Room room, Position other, Humanoid entity = null)
        {
            List<Position> potentialTiles = new List<Position>();

            for (int i = 0; i < Pathfinder.MovePoints.Length; i++)
                potentialTiles.Add(Copy() + Pathfinder.MovePoints[i]);


            potentialTiles.RemoveAll(x => !RoomTile.IsValidTile(room, entity, x));

            if (potentialTiles.Count > 0)
            {
                potentialTiles = potentialTiles.OrderBy(x => x.GetDistanceSquared(other)).ToList();
                return potentialTiles[0];
            }

            return null;
        }

        /// <summary>
        /// Get distance between two coordinates
        /// </summary>
        public int GetDistanceSquared(Position point)
        {
            int dx = X - point.X;
            int dy = Y - point.Y;

            return (int)Math.Sqrt((dx * dx) + (dy * dy));
        }

        public Position GetSquareInFront()
        {
            Position square = Copy();

            if (BodyRotation == 0)
            {
                square.Y--;
            }
            else if (BodyRotation == 1)
            {
                square.X++;
                square.Y--;
            }
            else if (BodyRotation == 2)
            {
                square.X++;
            }
            else if (BodyRotation == 3)
            {
                square.X++;
                square.Y++;
            }
            else if (BodyRotation == 4)
            {
                square.Y++;
            }
            else if (BodyRotation == 5)
            {
                square.X--;
                square.Y++;
            }
            else if (BodyRotation == 6)
            {
                square.X--;
            }
            else if (BodyRotation == 7)
            {
                square.X--;
                square.Y--;
            }

            return square;
        }


        public Position GetSquareBehind()
        {
            Position square = Copy();

            if (BodyRotation == 0)
            {
                square.Y++;
            }
            else if (BodyRotation == 1)
            {
                square.X--;
                square.Y++;
            }
            else if (BodyRotation == 2)
            {
                square.X--;
            }
            else if (BodyRotation == 3)
            {
                square.X--;
                square.Y--;
            }
            else if (BodyRotation == 4)
            {
                square.Y--;
            }
            else if (BodyRotation == 5)
            {
                square.X++;
                square.Y--;
            }
            else if (BodyRotation == 6)
            {
                square.X++;
            }
            else if (BodyRotation == 7)
            {
                square.X++;
                square.Y++;
            }

            return square;
        }

        public Position GetSquareRight()
        {
            Position square = Copy();

            if (BodyRotation == 0)
            {
                square.X++;
            }
            else if (BodyRotation == 1)
            {
                square.X++;
                square.Y++;
            }
            else if (BodyRotation == 2)
            {
                square.Y++;
            }
            else if (BodyRotation == 3)
            {
                square.X--;
                square.Y++;
            }
            else if (BodyRotation == 4)
            {
                square.X--;
            }
            else if (BodyRotation == 5)
            {
                square.X--;
                square.Y--;
            }
            else if (BodyRotation == 6)
            {
                square.Y--;
            }
            else if (BodyRotation == 7)
            {
                square.X++;
                square.Y--;
            }

            return square;
        }

        public Position GetSquareLeft()
        {
            Position square = Copy();

            if (BodyRotation == 0)
            {
                square.X--;
            }
            else if (BodyRotation == 1)
            {
                square.X--;
                square.Y--;
            }
            else if (BodyRotation == 2)
            {
                square.Y--;
            }
            else if (BodyRotation == 3)
            {
                square.X++;
                square.Y--;
            }
            else if (BodyRotation == 4)
            {
                square.X++;
            }
            else if (BodyRotation == 5)
            {
                square.X++;
                square.Y++;
            }
            else if (BodyRotation == 6)
            {
                square.Y++;
            }
            else if (BodyRotation == 7)
            {
                square.X--;
                square.Y++;
            }

            return square;
        }

        public static bool operator ==(Position one, Position two)
        {
            if (one is null && two is null)
                return true;

            if (one is Position && two is Position)
                return one.Equals(two);

            return false;
        }

        public static bool operator !=(Position one, Position two)
        {
            if (one is null && !(two is null))
                return true;

            if (!(one is null) && two is null)
                return true;

            if (one is Position && two is Position)
                return !one.Equals(two);

            return false;
        }

        public static Position operator +(Position one, Position two)
        {
            return one.Copy().Add(two);
        }

        public static Position operator -(Position one, Position two)
        {
            return new Position(one.X - two.X, one.Y - two.Y);
        }

        public Position Copy()
        {
            return new Position(X, Y, Z, HeadRotation, BodyRotation);
        }

        public override bool Equals(object obj)
        {
            if (obj == null)
                return false;

            var t = obj as Position;

            if (t == null)
                return false;

            if (X == t.X && Y == t.Y)
                return true;

            return false;
        }

        public override int GetHashCode()
        {
            return base.GetHashCode();
        }

        public override string ToString()
        {
            return $"[X: {X}, Y: {Y}, Z: {Z}, HeadRotation: {HeadRotation}, BodyRotation: {BodyRotation}]";
        }
    }
}

File: euclid/src/main/java/euclid/game/pathfinder/Position.java
package euclid.game.pathfinder;

import euclid.game.entity.Humanoid;
import euclid.game.room.Room;
import euclid.game.room.RoomTile;
import euclid.game.room.TileState;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class Position {

    private int x;
    private int y;
    private double z;
    private int bodyRotation;
    private int headRotation;

    public Position() {
        this(0, 0, 0);
    }

    public Position(int x, int y) {
        this.x = x;
        this.y = y;
        this.z = 0;
    }

    public Position(int x, int y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Position(int x, int y, double z, int headRotation, int bodyRotation) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.headRotation = headRotation;
        this.bodyRotation = bodyRotation;
    }

    public int getX() {
        return x;
    }

    public void setX(int x) {
        this.x = x;
    }

    public int getY() {
        return y;
    }

    public void setY(int y) {
        this.y = y;
    }

    public double getZ() {
        return z;
    }

    public void setZ(double z) {
        this.z = z;
    }

    public int getRotation() {
        return bodyRotation;
    }

    public void setRotation(int rotation) {
        this.bodyRotation = rotation;
        this.headRotation = rotation;
    }

    public int getBodyRotation() {
        return bodyRotation;
    }

    public void setBodyRotation(int bodyRotation) {
        this.bodyRotation = bodyRotation;
    }

    public int getHeadRotation() {
        return headRotation;
    }

    public void setHeadRotation(int headRotation) {
        this.headRotation = headRotation;
    }

    public RoomTile getTile(Room room) {
        if (room == null || room.getModel() == null || !room.getModel().isTile(this))
            return null;

        if (room.getModel().getTileStates()[this.x][this.y] == TileState.CLOSED)
            return null;

        return room.getMapping().getTiles()[this.x][this.y];
    }

    public boolean touches(Position position) {
        return getDistanceSquared(position) <= 2;
    }

    public Position add(Position other) {
        return new Position(other.getX() + x, other.getY() + y, other.getZ() + z);
    }

    public Position subtract(Position other) {
        return new Position(other.getX() - x, other.getY() - y, other.getZ() - z);
    }

    public Position closestTile(Room room, Position other, Humanoid entity) {
        List<Position> potentialTiles = new ArrayList<>();

        for (int i = 0; i < Pathfinder.MOVE_POINTS.length; i++)
            potentialTiles.add(copy().add(Pathfinder.MOVE_POINTS[i]));

        potentialTiles.removeIf(pos -> !RoomTile.isValidTile(room, entity, pos, false));

        if (!potentialTiles.isEmpty()) {
            potentialTiles.sort(Comparator.comparingInt(pos -> pos.getDistanceSquared(other)));
            return potentialTiles.get(0);
        }

        return null;
    }

    public int getDistanceSquared(Position point) {
        int dx = x - point.getX();
        int dy = y - point.getY();
        return (int) Math.sqrt((dx * dx) + (dy * dy));
    }

    public Position getSquareInFront() {
        Position square = copy();

        if (bodyRotation == 0) {
            square.y--;
        } else if (bodyRotation == 1) {
            square.x++;
            square.y--;
        } else if (bodyRotation == 2) {
            square.x++;
        } else if (bodyRotation == 3) {
            square.x++;
            square.y++;
        } else if (bodyRotation == 4) {
            square.y++;
        } else if (bodyRotation == 5) {
            square.x--;
            square.y++;
        } else if (bodyRotation == 6) {
            square.x--;
        } else if (bodyRotation == 7) {
            square.x--;
            square.y--;
        }

        return square;
    }

    public Position getSquareBehind() {
        Position square = copy();

        if (bodyRotation == 0) {
            square.y++;
        } else if (bodyRotation == 1) {
            square.x--;
            square.y++;
        } else if (bodyRotation == 2) {
            square.x--;
        } else if (bodyRotation == 3) {
            square.x--;
            square.y--;
        } else if (bodyRotation == 4) {
            square.y--;
        } else if (bodyRotation == 5) {
            square.x++;
            square.y--;
        } else if (bodyRotation == 6) {
            square.x++;
        } else if (bodyRotation == 7) {
            square.x++;
            square.y++;
        }

        return square;
    }

    public Position getSquareRight() {
        Position square = copy();

        if (bodyRotation == 0) {
            square.x++;
        } else if (bodyRotation == 1) {
            square.x++;
            square.y++;
        } else if (bodyRotation == 2) {
            square.y++;
        } else if (bodyRotation == 3) {
            square.x--;
            square.y++;
        } else if (bodyRotation == 4) {
            square.x--;
        } else if (bodyRotation == 5) {
            square.x--;
            square.y--;
        } else if (bodyRotation == 6) {
            square.y--;
        } else if (bodyRotation == 7) {
            square.x++;
            square.y--;
        }

        return square;
    }

    public Position getSquareLeft() {
        Position square = copy();

        if (bodyRotation == 0) {
            square.x--;
        } else if (bodyRotation == 1) {
            square.x--;
            square.y--;
        } else if (bodyRotation == 2) {
            square.y--;
        } else if (bodyRotation == 3) {
            square.x++;
            square.y--;
        } else if (bodyRotation == 4) {
            square.x++;
        } else if (bodyRotation == 5) {
            square.x++;
            square.y++;
        } else if (bodyRotation == 6) {
            square.y++;
        } else if (bodyRotation == 7) {
            square.x--;
            square.y++;
        }

        return square;
    }

    public Position copy() {
        return new Position(x, y, z, headRotation, bodyRotation);
    }

    public Position plus(Position other) {
        return copy().add(other);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Position other = (Position) obj;
        return x == other.x && y == other.y;
    }

    @Override
    public int hashCode() {
        return 31 * x + y;
    }

    @Override
    public String toString() {
        return String.format("[X: %d, Y: %d, Z: %f, HeadRotation: %d, BodyRotation: %d]", x, y, z, headRotation, bodyRotation);
    }
}
